### ISSUE TITLE
MulticlassJaccardIndex without reduction over N

### ISSUE THREAD

**Issue by MaKaNu** [Lightning-AI/metrics#1289]
## 🚀 Feature

The function `multilabel_jaccard_index` doesn't reduce over first dimension.

Small Example:

```python
pred = tensor([[0, 1],
               [0, 0]])

target = tensor([[[0, 1],
                  [0, 0]],
                 [[0, 1],
                  [0, 0]],
                 [[0, 1],
                  [0, 0]]])

target_number = target.shape[0]
expanded_pred = pred.repeat(target_number, 1, 1)


classes = 3

jaccard_idx = multiclass_jaccard_index(
            pred, target, num_classes=6, ignore_index=0, reduce=False)
```

This should not reduce the first dimension and sets the `average='none'`. 

### Motivation

I wrote this myself before with a bit different approach as I saw in your code. I recently discovered that IoU is simply jaccard index, so I was happy about finding this implementation. Sadly at the moment it doesn't fit my needing. I need the information to compare the single layer pred to multiple others.

### Pitch

At the moment `jaccard_idx` will be `tensor([0., 1., 0.])`, so I lost the information about the first dimension of the target. What I expected to see:

```python
> jaccard_idx
tensor( [[0., 1., 0.],
             [0., 1., 0.],
             [0., 1., 0.])
```
I played a bit around with the underscoremethods used by `multilabel_jaccard_index` and it seems that after `_multiclass_confusion_matrix_format` the first dimension is still intact but looses it in `_multiclass_confusion_matrix_update`. Further I discovered that it might be possible to create in the `_bincount` method an output with the size `(N, minlength)` and iterate over N additional will create the bins for the dim N individual. At that point I am not sure if `z.index_add_` has the same capabilities and if `jaccard_index_reduce` will be able to work with an `N x num_classes x num_classes` confusion matrix.

**Comment by SkafteNicki** [comment]
Hi @MaKaNu, thanks for raising this issue.
I am a bit confused about your usecase. In the first line of the issue you refer to `multilabel_jaccard_index` and the inputs in your example is clearly also `multilabel`. However, the function you are calling in the example is the `multiclass_jaccard_index` function. Therefore, are you trying to solve `multilabel` classification (all targets are [0,1] domain) or are you trying to solve `multiclass` classificatino (all targets are in [0, n_classes-1] domain).

**Comment by MaKaNu** [comment]
I use a bit different naming because my usecase is coming from the [boxsup paper](https://arxiv.org/abs/1503.01640). The algorithm compares candidates (predictions) against boundingboxes (target) and than multiply the Jaccard Index result with 1 if bbox and candidate are the same class. This means the result of Jaccard Index should not reduced along the first dim. As a secondary loss CrossEntropy is used which allows not to reduce the necessary candidates dimensions. 

I am not sure if multilabel serves my usage, but multiclass works as expected if target only has one candidate. as example:

```pyhton
target = tensor([[0, 1],
                  [0, 0]])
```

**Comment by SkafteNicki** [comment]
Alright, I am still not completely sure about the specifics of you usecase, but what I get is that you want to not reduce over two dimensions, which is not possible for our implementation of `jaccard_index` regardless of `multilabel` or `multiclass` setting. As it states in the [documentation](https://torchmetrics.readthedocs.io/en/stable/classification/jaccard_index.html#multiclassjaccardindex):
```
Additional dimension ... will be flattened into the batch dimension.
```
Therefore, your best way forward here would be to call it in some sequential way:
```python
torch.stack([multilabel_jaccard_index(p, t, num_labels=N, average=None) for p, t in zip(preds, target)])
```

**Comment by MaKaNu** [comment]
Okay that is relative close to my intermediate solution. In the meantime I have implemented two vectorized solution which might reduce the for loop overhead for my usecase.

**Comment by SkafteNicki** [comment]
Sounds good. We may in the future look into a more permanent solution for not reducing the metrics over the sample dimension, but for now we are closing this issue.
Thanks again for your interest in Torchmetrics, and sorry for not being able to provide the perfect solution for your usecase.
